src/gpl/src/mbff.h

```diff
@@ -0,0 +1,98 @@
+///////////////////////////////////////////////////////////////////////////////
+// BSD 3-Clause License
+//
+// Copyright (c) 2023, Precision Innovations Inc.
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+// * Redistributions of source code must retain the above copyright notice, this
+//   list of conditions and the following disclaimer.
+//
+// * Redistributions in binary form must reproduce the above copyright notice,
+//   this list of conditions and the following disclaimer in the documentation
+//   and/or other materials provided with the distribution.
+//
+// * Neither the name of the copyright holder nor the names of its
+//   contributors may be used to endorse or promote products derived from
+//   this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+// OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+///////////////////////////////////////////////////////////////////////////////
+
+#include <vector>
+#include "odb/db.h"
+
+#pragma once
+
+namespace gpl {
+
+struct Tray {
+    odb::Point pt;
+    std::vector<odb::Point> slots;
+    int cand[70];
```

Comment:
Prefer std::array<int, 70>. 

Also is there an explanation for the 70 number? Perhaps a constexpr variable with a name or comment would be helpful for readability

Reply:
Hi Ethan, thank you very much for the comment.
There are a maximum of 64 slots in a tray -- I just created the "70" as a hard limit. 

Perhaps it would be better to create a vector, and then resize it given the size of a tray?

Reply:
Generally, if there's a maximum value we would want the array size to represent exactly that number.

Otherwise a future reader might be confused why there are 70 elements in a tray, but the algorithm can only place 64. Could you summarize where the limit comes from? Is it an algorithm defined constraint, or just an arbitrary constant? Either way it would be good to introduce a variable name called MAX_SLOTS

The other consideration to make is that std::array<int, MAX_SLOTS>  and int[MAX_SLOTS] consume MAX_SLOTS worth of memory regardless if they're used or not. 

Sometimes that's good if we expect that slots will always be filled or close to filled. If we expect that most slots will be empty then a std::vector can be a better choice.

Reply:
We are working with a maximum of 64-bit trays, so there are a maximum of 64-bit slots. 
However, there are many trays created that only contain <= 2^5 slots (so, there would be a bunch of wasted memory if I just went with the array of size 64). 

In the most recent commit, I replaced the cand array with a vector to support resizing.

Reply:
Creating the cand vector and not adding a default value, even if the size is initialized, seems to increase runtime by ~1 minute (tested for #flops <= 5K, split-size = 500).

I initialized the cand vector to have a size of 1, and made the default value -1.
The vector then gets resized when we create a Tray that doesn't need as many slots. 

I'm not sure why this happens... 

Any suggestions?

Reply:
The resize penalty can be high. In this case it might be worth it to move to static arrays

Reply:
Hi, the issue doesn't persist if I just create a vector of size 1 with the value of -1. 
Should I just keep it as is (and then resize when trays are created), or should I create a static array of size 64 for every tray? 

The runtime differences are negligible with these methods.

Reply:
Is the number of flops known at allocation time? If so you should leave the vector empty and then use `std::vector.reserve()` before pushing elements.